app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt

```diff
@@ -68,6 +71,25 @@ class BrowserFragment : BaseBrowserFragment(), UserInteractionHandler {
         val components = context.components
 
         return super.initializeUI(view)?.also {
+            val readerModeAction =
+                BrowserToolbar.ToggleButton(
+                    image = ContextCompat.getDrawable(requireContext(), R.drawable.ic_readermode)!!,
+                    imageSelected = ContextCompat.getDrawable(requireContext(), R.drawable.ic_readermode_selected)!!,
+                    contentDescription = requireContext().getString(R.string.browser_menu_read),
+                    contentDescriptionSelected = requireContext().getString(R.string.browser_menu_read_close),
+                    visible = {
+                        getSessionById()?.let {
+                            activity?.components?.core?.store?.state?.findTab(it.id)?.readerState?.readerable
```

Comment:
Unfortunately, this isn't sufficient here. Whether or not a page is readerable is determined once the page is loaded and once our built-in web extension (using readability.js) determined that. This may or may not have happened at this point.

In general, I much prefer this solution using the toolbar, as the user will be notified that a page is readerable instead of having to check the menu. Fenix invalidates the menu for that exact reason to make sure it's up-to-date once we know if the page is "readerable": https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt#L82

In R-B and Sample Browser, which also uses a page action in the toolbar we do this: https://github.com/mozilla-mobile/android-components/blob/master/samples/browser/src/main/java/org/mozilla/samples/browser/integration/ReaderViewIntegration.kt#L58

This is important, otherwise we may miss to display the button, although the page is readerable.

Reply:
So it should be sufficient to make sure the button visibility is updated when this callback is invoked: https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/browser/BrowserFragment.kt#L78

Reply:
You can see the problem in action on cnbc.com, or another slower site. Click on an article that's readerable (right now "NFL condemns racism..."), the button doesn't appear. Refresh and it does.

Reply:
I've updated this and it should be fixed now 😄